lib/player.js

```diff
@@ -753,25 +756,50 @@ shaka.Player.prototype.isBuffering = function() {
  * @export
  */
 shaka.Player.prototype.unload = function() {
-  if (this.destroyed_) return Promise.resolve();
+  if (this.unloadPromise_) {
+    // An unload is currently in progress. Do not initiate another one.
+    return this.unloadPromise_;
+  }
+
+  if (this.destroyed_) {
+    // No unloading necessary.
+    return Promise.resolve();
+  }
+
   this.dispatchEvent(new shaka.util.FakeEvent('unloading'));
 
   if (this.loadChain_) {
     // A load is in progress.  Cancel it, then reset the streaming system.
     var interrupt = new shaka.util.Error(
         shaka.util.Error.Category.PLAYER,
         shaka.util.Error.Code.LOAD_INTERRUPTED);
-    return this.loadChain_.cancel(interrupt)
-        .then(this.resetStreaming_.bind(this));
+    return this.createUnloadPromise_(this.loadChain_.cancel(interrupt)
+        .then(this.resetStreaming_.bind(this)));
   } else {
-    // No loads or unloads are in progress.
-    // Just reset the streaming system if needed.
-    return this.resetStreaming_();
+    // No loads are in progress. Just reset the streaming system if needed.
+    return this.createUnloadPromise_(this.resetStreaming_());
   }
 };
 
 
 /**
+ * Create an unload promise and store it on the player until unloading
+ * is complete.
+ * @private
+ * @param {!Promise} unloadProcess Unloading steps to perform before destroying
+ *     the unload promise.
+ * @return {!Promise}
+ */
+shaka.Player.prototype.createUnloadPromise_ = function(unloadProcess) {
```

Comment:
Since you're passing in an unresolved Promise chain, and to keep terminology consistent with other parts of this file, change "unloadProcess" to "unloadChain" or "unloadPromiseChain".